win_perf_counters - measurement gaps

### Relevant telegraf.conf

```toml
# Get all Windows process performance counters
[[inputs.win_perf_counters]]
  alias = "perf_counters_test"
  UseWildcardsExpansion = true
  interval = "30s"
  PrintValid=true

  [[inputs.win_perf_counters.object]]
    ObjectName = "Process"
    Instances = ["*"]
    Counters = ["*"]
    Measurement = "win_process_test"
```


### Logs from Telegraf

```text
## usual output
2023-10-12T06:51:00Z I! [inputs.win_perf_counters::perf_counters_test] Valid: \\myhostname\Process(Idle)\% Processor Time
2023-10-12T06:51:00Z I! [inputs.win_perf_counters::perf_counters_test] Valid: \\myhostname\Process(Idle)\% User Time
2023-10-12T06:51:00Z I! [inputs.win_perf_counters::perf_counters_test] Valid: \\myhostname\Process(Idle)\% Privileged Time
2023-10-12T06:51:00Z I! [inputs.win_perf_counters::perf_counters_test] Valid: \\myhostname\Process(Idle)\Virtual Bytes Peak
2023-10-12T06:51:00Z I! [inputs.win_perf_counters::perf_counters_test] Valid: \\myhostname\Process(Idle)\Virtual Bytes
...
2023-10-12T06:51:01Z D! [inputs.win_perf_counters::perf_counters_test] gathering from localhost
2023-10-12T06:51:01Z D! [inputs.win_perf_counters::perf_counters_test] gathering from localhost finished in %!f(time.Duration=10253900)s


## When missing:

(No "Valid:...")
2023-10-12T06:49:31Z D! [inputs.win_perf_counters::perf_counters_test] gathering from localhost
2023-10-12T06:49:31Z D! [inputs.win_perf_counters::perf_counters_test] gathering from localhost finished in %!f(time.Duration=000)s
(Again, no "Valid:...")
2023-10-12T06:50:00Z D! [inputs.win_perf_counters::perf_counters_test] gathering from localhost
2023-10-12T06:50:00Z D! [inputs.win_perf_counters::perf_counters_test] gathering from localhost finished in %!f(time.Duration=000)s
(And again, no "Valid:...")
2023-10-12T06:50:30Z D! [inputs.win_perf_counters::perf_counters_test] gathering from localhost
2023-10-12T06:50:30Z D! [inputs.win_perf_counters::perf_counters_test] gathering from localhost finished in %!f(time.Duration=000)s
...
and then it's back to working again.
```


### System info

Telegraf 1.27.3 (git: HEAD@afcf0133), Windows Server 2022

### Docker

_No response_

### Steps to reproduce

1. Use the above config
2. Observe 4-8 gaps a week.




### Expected behavior

No gaps

### Actual behavior

Gaps

### Additional info

It always seems to fail thrice in a row, I suspect it's because how the "last time refreshed" is recorded.

1. 6:49:30 - plugin starts, has to refresh
2. 6:49:31 - refreshed and recorded "refreshed - 6:49:31", then doing a gather.
3. somehow there are maybe no counters? 
I believe this for loop
https://github.com/influxdata/telegraf/blob/master/plugins/inputs/win_perf_counters/win_perf_counters.go#L245
is going empty, not producing any measurements and not logging any "Valid:..."s.
Gathering nothing and not looping at all takes time.Duration=000

4. 6:50:00 - No new refresh, and as the refresh at 6:49:30 didn't produce any counters, again no measurements

5. 6:50:30 - The last refresh is not 60s ago, yet, as it is 6:50:30 and last refresh was recorded at 6:49:31.
-> still no measurements.

So:

It seems apparent why there's always a gap of three and roughly where to look for the error, but I can't find it, lest it be windows itself returning an empty process list and no error. (not even PDH_NO_DATA)

---
![grafik](https://github.com/influxdata/telegraf/assets/131003185/5c465b21-a1f1-43b4-8650-7047a5c06054)

We have had this problem for months and many telegraf-versions now. It's a "sometimes" problem (which is always bad, it's a bad look for reliability), so we tried gathering information and solving it on our own, first.

I'd like to, but it seems you didn't test compile your changes before committing?
```
# github.com/influxdata/telegraf/plugins/inputs/win_perf_counters
plugins/inputs/win_perf_counters/win_perf_counters.go:591:70: cannot use math.MaxUint32 (untyped int constant 4294967295) as int value in argument to fmt.Errorf (overflows)
make[1]: *** [Makefile:277: build/windows-386/telegraf.exe] Error 1
make: *** [Makefile:343: windows_i386.zip] Error 2
```

--- a/plugins/inputs/win_perf_counters/README.md
+++ b/plugins/inputs/win_perf_counters/README.md
@@ -345,6 +345,10 @@ See the [CONFIGURATION.md][CONFIGURATION.md] for more details.
   ## e.g. IgnoredErrors = ["PDH_NO_DATA"]
   # IgnoredErrors = []
 
+  ## Maximum size of the buffer for values returned by the API
+  ## Increase this value if you experience "buffer limit reached" errors.
+  # MaxBufferSize = "4MiB"
+
   ## NOTE: Due to the way TOML is parsed, tables must be at the END of the
   ## plugin definition, otherwise additional config options are read as part of
   ## the table

--- a/plugins/inputs/win_perf_counters/performance_query.go
+++ b/plugins/inputs/win_perf_counters/performance_query.go
@@ -10,6 +10,11 @@ import (
 	"unsafe"
 )
 
+// Initial buffer size for return buffers
+const initialBufferSize = uint32(1024) // 1kB
+
+var errBufferLimitReached = errors.New("buffer limit reached")
+
 // CounterValue is abstraction for PdhFmtCountervalueItemDouble
 type CounterValue struct {
 	InstanceName string
@@ -36,7 +41,7 @@ type PerformanceQuery interface {
 }
 
 type PerformanceQueryCreator interface {
-	NewPerformanceQuery(string) PerformanceQuery
+	NewPerformanceQuery(string, uint32) PerformanceQuery
 }
 
 // PdhError represents error returned from Performance Counters API
@@ -58,14 +63,14 @@ func NewPdhError(code uint32) error {
 
 // PerformanceQueryImpl is implementation of PerformanceQuery interface, which calls phd.dll functions
 type PerformanceQueryImpl struct {
-	query pdhQueryHandle
+	maxBufferSize uint32
+	query         pdhQueryHandle
 }
 
-type PerformanceQueryCreatorImpl struct {
-}
+type PerformanceQueryCreatorImpl struct{}
 
-func (m PerformanceQueryCreatorImpl) NewPerformanceQuery(string) PerformanceQuery {
-	return &PerformanceQueryImpl{}
+func (m PerformanceQueryCreatorImpl) NewPerformanceQuery(_ string, maxBufferSize uint32) PerformanceQuery {
+	return &PerformanceQueryImpl{maxBufferSize: maxBufferSize}
 }
 
 // Open creates a new counterPath that is used to manage the collection of performance data.
@@ -124,64 +129,82 @@ func (m *PerformanceQueryImpl) AddEnglishCounterToQuery(counterPath string) (pdh
 
 // GetCounterPath return counter information for given handle
 func (m *PerformanceQueryImpl) GetCounterPath(counterHandle pdhCounterHandle) (string, error) {
-	var bufSize uint32
-	var buff []byte
-	var ret uint32
-	if ret = PdhGetCounterInfo(counterHandle, 0, &bufSize, nil); ret == PdhMoreData {
-		buff = make([]byte, bufSize)
-		bufSize = uint32(len(buff))
-		if ret = PdhGetCounterInfo(counterHandle, 0, &bufSize, &buff[0]); ret == ErrorSuccess {
-			ci := (*PdhCounterInfo)(unsafe.Pointer(&buff[0])) //nolint:gosec // G103: Valid use of unsafe call to create PDH_COUNTER_INFO
+	for buflen := initialBufferSize; buflen <= m.maxBufferSize; buflen *= 2 {
+		buf := make([]byte, buflen)
+
+		// Get the info with the current buffer size
+		size := buflen
+		ret := PdhGetCounterInfo(counterHandle, 0, &size, &buf[0])
+		if ret == ErrorSuccess {
+			ci := (*PdhCounterInfo)(unsafe.Pointer(&buf[0])) //nolint:gosec // G103: Valid use of unsafe call to create PDH_COUNTER_INFO
 			return UTF16PtrToString(ci.SzFullPath), nil
 		}
+
+		// Use the size as a hint if it exceeds the current buffer size
+		if size > buflen {
+			buflen = size
+		}
+
+		// We got a non-recoverable error so exit here
+		if ret != PdhMoreData {
+			return "", NewPdhError(ret)
+		}
 	}
-	return "", NewPdhError(ret)
+
+	return "", errBufferLimitReached
 }
 
 // ExpandWildCardPath  examines local computer and returns those counter paths that match the given counter path which contains wildcard characters.
 func (m *PerformanceQueryImpl) ExpandWildCardPath(counterPath string) ([]string, error) {
-	var bufSize uint32
-	var buff []uint16
-	var ret uint32
+	for buflen := initialBufferSize; buflen <= m.maxBufferSize; buflen *= 2 {
+		buf := make([]uint16, buflen)
 
-	if ret = PdhExpandWildCardPath(counterPath, nil, &bufSize); ret == PdhMoreData {
-		buff = make([]uint16, bufSize)
-		bufSize = uint32(len(buff))
-		ret = PdhExpandWildCardPath(counterPath, &buff[0], &bufSize)
+		// Get the info with the current buffer size
+		size := buflen
+		ret := PdhExpandWildCardPath(counterPath, &buf[0], &size)
 		if ret == ErrorSuccess {
-			list := UTF16ToStringArray(buff)
-			return list, nil
+			return UTF16ToStringArray(buf), nil
+		}
+
+		// Use the size as a hint if it exceeds the current buffer size
+		if size > buflen {
+			buflen = size
+		}
+
+		// We got a non-recoverable error so exit here
+		if ret != PdhMoreData {
+			return nil, NewPdhError(ret)
 		}
 	}
-	return nil, NewPdhError(ret)
+
+	return nil, errBufferLimitReached
 }
 
 // GetFormattedCounterValueDouble computes a displayable value for the specified counter
 func (m *PerformanceQueryImpl) GetFormattedCounterValueDouble(hCounter pdhCounterHandle) (float64, error) {
 	var counterType uint32
 	var value PdhFmtCountervalueDouble
-	var ret uint32
 
-	if ret = PdhGetFormattedCounterValueDouble(hCounter, &counterType, &value); ret == ErrorSuccess {
-		if value.CStatus == PdhCstatusValidData || value.CStatus == PdhCstatusNewData {
-			return value.DoubleValue, nil
-		}
-		return 0, NewPdhError(value.CStatus)
+	if ret := PdhGetFormattedCounterValueDouble(hCounter, &counterType, &value); ret != ErrorSuccess {
+		return 0, NewPdhError(ret)
 	}
-	return 0, NewPdhError(ret)
+	if value.CStatus == PdhCstatusValidData || value.CStatus == PdhCstatusNewData {
+		return value.DoubleValue, nil
+	}
+	return 0, NewPdhError(value.CStatus)
 }
 
 func (m *PerformanceQueryImpl) GetFormattedCounterArrayDouble(hCounter pdhCounterHandle) ([]CounterValue, error) {
-	var buffSize uint32
-	var itemCount uint32
-	var ret uint32
+	for buflen := initialBufferSize; buflen <= m.maxBufferSize; buflen *= 2 {
+		buf := make([]byte, buflen)
 
-	if ret = PdhGetFormattedCounterArrayDouble(hCounter, &buffSize, &itemCount, nil); ret == PdhMoreData {
-		buff := make([]byte, buffSize)
-
-		if ret = PdhGetFormattedCounterArrayDouble(hCounter, &buffSize, &itemCount, &buff[0]); ret == ErrorSuccess {
+		// Get the info with the current buffer size
+		var itemCount uint32
+		size := buflen
+		ret := PdhGetFormattedCounterArrayDouble(hCounter, &size, &itemCount, &buf[0])
+		if ret == ErrorSuccess {
 			//nolint:gosec // G103: Valid use of unsafe call to create PDH_FMT_COUNTERVALUE_ITEM_DOUBLE
-			items := (*[1 << 20]PdhFmtCountervalueItemDouble)(unsafe.Pointer(&buff[0]))[:itemCount]
+			items := (*[1 << 20]PdhFmtCountervalueItemDouble)(unsafe.Pointer(&buf[0]))[:itemCount]
 			values := make([]CounterValue, 0, itemCount)
 			for _, item := range items {
 				if item.FmtValue.CStatus == PdhCstatusValidData || item.FmtValue.CStatus == PdhCstatusNewData {
@@ -191,21 +214,32 @@ func (m *PerformanceQueryImpl) GetFormattedCounterArrayDouble(hCounter pdhCounte
 			}
 			return values, nil
 		}
+
+		// Use the size as a hint if it exceeds the current buffer size
+		if size > buflen {
+			buflen = size
+		}
+
+		// We got a non-recoverable error so exit here
+		if ret != PdhMoreData {
+			return nil, NewPdhError(ret)
+		}
 	}
-	return nil, NewPdhError(ret)
+
+	return nil, errBufferLimitReached
 }
 
 func (m *PerformanceQueryImpl) GetRawCounterArray(hCounter pdhCounterHandle) ([]CounterValue, error) {
-	var buffSize uint32
-	var itemCount uint32
-	var ret uint32
+	for buflen := initialBufferSize; buflen <= m.maxBufferSize; buflen *= 2 {
+		buf := make([]byte, buflen)
 
-	if ret = PdhGetRawCounterArray(hCounter, &buffSize, &itemCount, nil); ret == PdhMoreData {
-		buff := make([]byte, buffSize)
-
-		if ret = PdhGetRawCounterArray(hCounter, &buffSize, &itemCount, &buff[0]); ret == ErrorSuccess {
+		// Get the info with the current buffer size
+		var itemCount uint32
+		size := buflen
+		ret := PdhGetRawCounterArray(hCounter, &size, &itemCount, &buf[0])
+		if ret == ErrorSuccess {
 			//nolint:gosec // G103: Valid use of unsafe call to create PDH_RAW_COUNTER_ITEM
-			items := (*[1 << 20]PdhRawCounterItem)(unsafe.Pointer(&buff[0]))[:itemCount]
+			items := (*[1 << 20]PdhRawCounterItem)(unsafe.Pointer(&buf[0]))[:itemCount]
 			values := make([]CounterValue, 0, itemCount)
 			for _, item := range items {
 				if item.RawValue.CStatus == PdhCstatusValidData || item.RawValue.CStatus == PdhCstatusNewData {
@@ -215,8 +249,19 @@ func (m *PerformanceQueryImpl) GetRawCounterArray(hCounter pdhCounterHandle) ([]
 			}
 			return values, nil
 		}
+
+		// Use the size as a hint if it exceeds the current buffer size
+		if size > buflen {
+			buflen = size
+		}
+
+		// We got a non-recoverable error so exit here
+		if ret != PdhMoreData {
+			return nil, NewPdhError(ret)
+		}
 	}
-	return nil, NewPdhError(ret)
+
+	return nil, errBufferLimitReached
 }
 
 func (m *PerformanceQueryImpl) CollectData() error {

--- a/plugins/inputs/win_perf_counters/sample.conf
+++ b/plugins/inputs/win_perf_counters/sample.conf
@@ -41,6 +41,10 @@
   ## e.g. IgnoredErrors = ["PDH_NO_DATA"]
   # IgnoredErrors = []
 
+  ## Maximum size of the buffer for values returned by the API
+  ## Increase this value if you experience "buffer limit reached" errors.
+  # MaxBufferSize = "4MiB"
+
   ## NOTE: Due to the way TOML is parsed, tables must be at the END of the
   ## plugin definition, otherwise additional config options are read as part of
   ## the table

--- a/plugins/inputs/win_perf_counters/win_perf_counters.go
+++ b/plugins/inputs/win_perf_counters/win_perf_counters.go
@@ -21,7 +21,7 @@ import (
 //go:embed sample.conf
 var sampleConfig string
 
-var defaultMaxBufferSize = config.Size(100 * 1024 * 1024)
+var defaultMaxBufferSize = config.Size(4 * 1024 * 1024)
 
 type WinPerfCounters struct {
 	PrintValid                 bool `toml:"PrintValid"`

--- a/plugins/inputs/win_perf_counters/win_perf_counters.go
+++ b/plugins/inputs/win_perf_counters/win_perf_counters.go
@@ -21,7 +21,7 @@ import (
 //go:embed sample.conf
 var sampleConfig string
 
-var defaultMaxBufferSize = config.Size(4 * 1024 * 1024)
+var defaultMaxBufferSize = config.Size(100 * 1024 * 1024)
 
 type WinPerfCounters struct {
 	PrintValid                 bool `toml:"PrintValid"`
@@ -588,7 +588,7 @@ func (m *WinPerfCounters) Init() error {
 		return fmt.Errorf("maximum buffer size should at least be %d", 2*initialBufferSize)
 	}
 	if m.MaxBufferSize > math.MaxUint32 {
-		return fmt.Errorf("maximum buffer size should be smaller than %d", math.MaxUint32)
+		return fmt.Errorf("maximum buffer size should be smaller than %d", uint32(math.MaxUint32))
 	}
 
 	if m.UseWildcardsExpansion && !m.LocalizeWildcardsExpansion {

--- a/plugins/inputs/win_perf_counters/win_perf_counters.go
+++ b/plugins/inputs/win_perf_counters/win_perf_counters.go
@@ -7,6 +7,7 @@ import (
 	_ "embed"
 	"errors"
 	"fmt"
+	"math"
 	"os"
 	"strings"
 	"sync"
@@ -20,6 +21,8 @@ import (
 //go:embed sample.conf
 var sampleConfig string
 
+var defaultMaxBufferSize = config.Size(100 * 1024 * 1024)
+
 type WinPerfCounters struct {
 	PrintValid                 bool `toml:"PrintValid"`
 	PreVistaSupport            bool `toml:"PreVistaSupport" deprecated:"1.7.0;determined dynamically"`
@@ -29,6 +32,7 @@ type WinPerfCounters struct {
 	UseWildcardsExpansion      bool
 	LocalizeWildcardsExpansion bool
 	IgnoredErrors              []string `toml:"IgnoredErrors"`
+	MaxBufferSize              config.Size
 	Sources                    []string
 
 	Log telegraf.Logger
@@ -207,7 +211,7 @@ func (m *WinPerfCounters) AddItem(counterPath, computer, objectName, instance, c
 	if !ok {
 		hostCounter = &hostCountersInfo{computer: computer, tag: sourceTag}
 		m.hostCounters[computer] = hostCounter
-		hostCounter.query = m.queryCreator.NewPerformanceQuery(computer)
+		hostCounter.query = m.queryCreator.NewPerformanceQuery(computer, uint32(m.MaxBufferSize))
 		if err := hostCounter.query.Open(); err != nil {
 			return err
 		}
@@ -579,9 +583,16 @@ func isKnownCounterDataError(err error) bool {
 }
 
 func (m *WinPerfCounters) Init() error {
+	// Check the buffer size
+	if m.MaxBufferSize < config.Size(initialBufferSize) {
+		return fmt.Errorf("maximum buffer size should at least be %d", 2*initialBufferSize)
+	}
+	if m.MaxBufferSize > math.MaxUint32 {
+		return fmt.Errorf("maximum buffer size should be smaller than %d", math.MaxUint32)
+	}
+
 	if m.UseWildcardsExpansion && !m.LocalizeWildcardsExpansion {
 		// Counters must not have wildcards with this option
-
 		found := false
 		wildcards := []string{"*", "?"}
 
@@ -614,6 +625,7 @@ func init() {
 		return &WinPerfCounters{
 			CountersRefreshInterval:    config.Duration(time.Second * 60),
 			LocalizeWildcardsExpansion: true,
+			MaxBufferSize:              defaultMaxBufferSize,
 			queryCreator:               &PerformanceQueryCreatorImpl{},
 		}
 	})

--- a/plugins/inputs/win_perf_counters/win_perf_counters_integration_test.go
+++ b/plugins/inputs/win_perf_counters/win_perf_counters_integration_test.go
@@ -18,7 +18,7 @@ func TestWinPerformanceQueryImplIntegration(t *testing.T) {
 	if testing.Short() {
 		t.Skip("Skipping integration test in short mode")
 	}
-	query := &PerformanceQueryImpl{}
+	query := &PerformanceQueryImpl{maxBufferSize: uint32(defaultMaxBufferSize)}
 
 	err := query.Close()
 	require.Error(t, err, "uninitialized query must return errors")
@@ -62,11 +62,11 @@ func TestWinPerformanceQueryImplIntegration(t *testing.T) {
 
 	fcounter, err := query.GetFormattedCounterValueDouble(hCounter)
 	require.NoError(t, err)
-	require.Greater(t, fcounter, 0)
+	require.Greater(t, fcounter, float64(0))
 
 	rcounter, err := query.GetRawCounterValue(hCounter)
 	require.NoError(t, err)
-	require.Greater(t, rcounter, 10000000)
+	require.Greater(t, rcounter, int64(10000000))
 
 	now := time.Now()
 	mtime, err := query.CollectDataWithTime()
@@ -131,10 +131,11 @@ func TestWinPerfCountersConfigGet1Integration(t *testing.T) {
 	}}
 
 	m := WinPerfCounters{
-		PrintValid:   false,
-		Object:       perfObjects,
-		queryCreator: &PerformanceQueryCreatorImpl{},
-		Log:          testutil.Logger{},
+		PrintValid:    false,
+		Object:        perfObjects,
+		MaxBufferSize: defaultMaxBufferSize,
+		Log:           testutil.Logger{},
+		queryCreator:  &PerformanceQueryCreatorImpl{},
 	}
 
 	require.NoError(t, m.ParseConfig())
@@ -158,10 +159,11 @@ func TestWinPerfCountersConfigGet2Integration(t *testing.T) {
 	}}
 
 	m := WinPerfCounters{
-		PrintValid:   false,
-		Object:       perfObjects,
-		queryCreator: &PerformanceQueryCreatorImpl{},
-		Log:          testutil.Logger{},
+		PrintValid:    false,
+		Object:        perfObjects,
+		MaxBufferSize: defaultMaxBufferSize,
+		Log:           testutil.Logger{},
+		queryCreator:  &PerformanceQueryCreatorImpl{},
 	}
 
 	require.NoError(t, m.ParseConfig())
@@ -200,10 +202,11 @@ func TestWinPerfCountersConfigGet3Integration(t *testing.T) {
 	}}
 
 	m := WinPerfCounters{
-		PrintValid:   false,
-		Object:       perfObjects,
-		queryCreator: &PerformanceQueryCreatorImpl{},
-		Log:          testutil.Logger{},
+		PrintValid:    false,
+		Object:        perfObjects,
+		MaxBufferSize: defaultMaxBufferSize,
+		Log:           testutil.Logger{},
+		queryCreator:  &PerformanceQueryCreatorImpl{},
 	}
 
 	require.NoError(t, m.ParseConfig())
@@ -240,10 +243,11 @@ func TestWinPerfCountersConfigGet4Integration(t *testing.T) {
 	}}
 
 	m := WinPerfCounters{
-		PrintValid:   false,
-		Object:       perfObjects,
-		queryCreator: &PerformanceQueryCreatorImpl{},
-		Log:          testutil.Logger{},
+		PrintValid:    false,
+		Object:        perfObjects,
+		MaxBufferSize: defaultMaxBufferSize,
+		Log:           testutil.Logger{},
+		queryCreator:  &PerformanceQueryCreatorImpl{},
 	}
 
 	require.NoError(t, m.ParseConfig())
@@ -280,10 +284,11 @@ func TestWinPerfCountersConfigGet5Integration(t *testing.T) {
 	}}
 
 	m := WinPerfCounters{
-		PrintValid:   false,
-		Object:       perfObjects,
-		queryCreator: &PerformanceQueryCreatorImpl{},
-		Log:          testutil.Logger{},
+		PrintValid:    false,
+		Object:        perfObjects,
+		MaxBufferSize: defaultMaxBufferSize,
+		Log:           testutil.Logger{},
+		queryCreator:  &PerformanceQueryCreatorImpl{},
 	}
 
 	require.NoError(t, m.ParseConfig())
@@ -320,10 +325,11 @@ func TestWinPerfCountersConfigGet6Integration(t *testing.T) {
 	}}
 
 	m := WinPerfCounters{
-		PrintValid:   false,
-		Object:       perfObjects,
-		queryCreator: &PerformanceQueryCreatorImpl{},
-		Log:          testutil.Logger{},
+		PrintValid:    false,
+		Object:        perfObjects,
+		MaxBufferSize: defaultMaxBufferSize,
+		Log:           testutil.Logger{},
+		queryCreator:  &PerformanceQueryCreatorImpl{},
 	}
 
 	require.NoError(t, m.ParseConfig())
@@ -347,10 +353,11 @@ func TestWinPerfCountersConfigGet7Integration(t *testing.T) {
 	}}
 
 	m := WinPerfCounters{
-		PrintValid:   false,
-		Object:       perfObjects,
-		queryCreator: &PerformanceQueryCreatorImpl{},
-		Log:          testutil.Logger{},
+		PrintValid:    false,
+		Object:        perfObjects,
+		MaxBufferSize: defaultMaxBufferSize,
+		Log:           testutil.Logger{},
+		queryCreator:  &PerformanceQueryCreatorImpl{},
 	}
 
 	require.NoError(t, m.ParseConfig())
@@ -387,10 +394,11 @@ func TestWinPerfCountersConfigError1Integration(t *testing.T) {
 	}}
 
 	m := WinPerfCounters{
-		PrintValid:   false,
-		Object:       perfObjects,
-		queryCreator: &PerformanceQueryCreatorImpl{},
-		Log:          testutil.Logger{},
+		PrintValid:    false,
+		Object:        perfObjects,
+		MaxBufferSize: defaultMaxBufferSize,
+		Log:           testutil.Logger{},
+		queryCreator:  &PerformanceQueryCreatorImpl{},
 	}
 
 	require.Error(t, m.ParseConfig())
@@ -414,10 +422,11 @@ func TestWinPerfCountersConfigError2Integration(t *testing.T) {
 	}}
 
 	m := WinPerfCounters{
-		PrintValid:   false,
-		Object:       perfObjects,
-		queryCreator: &PerformanceQueryCreatorImpl{},
-		Log:          testutil.Logger{},
+		PrintValid:    false,
+		Object:        perfObjects,
+		MaxBufferSize: defaultMaxBufferSize,
+		Log:           testutil.Logger{},
+		queryCreator:  &PerformanceQueryCreatorImpl{},
 	}
 
 	require.NoError(t, m.ParseConfig())
@@ -443,10 +452,11 @@ func TestWinPerfCountersConfigError3Integration(t *testing.T) {
 	}}
 
 	m := WinPerfCounters{
-		PrintValid:   false,
-		Object:       perfObjects,
-		queryCreator: &PerformanceQueryCreatorImpl{},
-		Log:          testutil.Logger{},
+		PrintValid:    false,
+		Object:        perfObjects,
+		MaxBufferSize: defaultMaxBufferSize,
+		Log:           testutil.Logger{},
+		queryCreator:  &PerformanceQueryCreatorImpl{},
 	}
 
 	require.Error(t, m.ParseConfig())
@@ -470,10 +480,11 @@ func TestWinPerfCountersCollect1Integration(t *testing.T) {
 	}}
 
 	m := WinPerfCounters{
-		PrintValid:   false,
-		Object:       perfObjects,
-		queryCreator: &PerformanceQueryCreatorImpl{},
-		Log:          testutil.Logger{},
+		PrintValid:    false,
+		Object:        perfObjects,
+		MaxBufferSize: defaultMaxBufferSize,
+		Log:           testutil.Logger{},
+		queryCreator:  &PerformanceQueryCreatorImpl{},
 	}
 
 	var acc testutil.Accumulator
@@ -510,9 +521,10 @@ func TestWinPerfCountersCollect2Integration(t *testing.T) {
 		PrintValid:            false,
 		UsePerfCounterTime:    true,
 		Object:                perfObjects,
-		queryCreator:          &PerformanceQueryCreatorImpl{},
 		UseWildcardsExpansion: true,
+		MaxBufferSize:         defaultMaxBufferSize,
 		Log:                   testutil.Logger{},
+		queryCreator:          &PerformanceQueryCreatorImpl{},
 	}
 
 	var acc testutil.Accumulator
@@ -550,9 +562,10 @@ func TestWinPerfCountersCollectRawIntegration(t *testing.T) {
 	m := WinPerfCounters{
 		PrintValid:            false,
 		Object:                perfObjects,
-		queryCreator:          &PerformanceQueryCreatorImpl{},
 		UseWildcardsExpansion: true,
+		MaxBufferSize:         defaultMaxBufferSize,
 		Log:                   testutil.Logger{},
+		queryCreator:          &PerformanceQueryCreatorImpl{},
 	}
 	var acc testutil.Accumulator
 	require.NoError(t, m.Gather(&acc))
@@ -566,17 +579,18 @@ func TestWinPerfCountersCollectRawIntegration(t *testing.T) {
 		val, ok := metric.Fields[expectedCounter]
 		require.True(t, ok, "Expected presence of %s field", expectedCounter)
 		valInt64, ok := val.(int64)
-		require.True(t, ok, fmt.Sprintf("Expected int64, got %T", val))
-		require.Greater(t, valInt64, 0, fmt.Sprintf("Expected > 0, got %d, for %#v", valInt64, metric))
+		require.Truef(t, ok, "Expected int64, got %T", val)
+		require.Greaterf(t, valInt64, int64(0), "Expected > 0, got %d, for %#v", valInt64, metric)
 	}
 
 	// Test *Array way
 	m = WinPerfCounters{
 		PrintValid:            false,
 		Object:                perfObjects,
-		queryCreator:          &PerformanceQueryCreatorImpl{},
 		UseWildcardsExpansion: false,
+		MaxBufferSize:         defaultMaxBufferSize,
 		Log:                   testutil.Logger{},
+		queryCreator:          &PerformanceQueryCreatorImpl{},
 	}
 	var acc2 testutil.Accumulator
 	require.NoError(t, m.Gather(&acc))
@@ -589,7 +603,7 @@ func TestWinPerfCountersCollectRawIntegration(t *testing.T) {
 		val, ok := metric.Fields[expectedCounter]
 		require.True(t, ok, "Expected presence of %s field", expectedCounter)
 		valInt64, ok := val.(int64)
-		require.True(t, ok, fmt.Sprintf("Expected int64, got %T", val))
-		require.Greater(t, valInt64, 0, fmt.Sprintf("Expected > 0, got %d, for %#v", valInt64, metric))
+		require.Truef(t, ok, "Expected int64, got %T", val)
+		require.Greaterf(t, valInt64, int64(0), "Expected > 0, got %d, for %#v", valInt64, metric)
 	}
 }

--- a/plugins/inputs/win_perf_counters/win_perf_counters_test.go
+++ b/plugins/inputs/win_perf_counters/win_perf_counters_test.go
@@ -214,7 +214,7 @@ type FakePerformanceQueryCreator struct {
 	fakeQueries map[string]*FakePerformanceQuery
 }
 
-func (m FakePerformanceQueryCreator) NewPerformanceQuery(computer string) PerformanceQuery {
+func (m FakePerformanceQueryCreator) NewPerformanceQuery(computer string, _ uint32) PerformanceQuery {
 	var ret PerformanceQuery
 	var ok bool
 	if ret, ok = m.fakeQueries[computer]; !ok {
@@ -2043,6 +2043,7 @@ func TestLocalizeWildcardsExpansion(t *testing.T) {
 			[]string{"_Total"}, []string{counter}, true, false, false),
 		LocalizeWildcardsExpansion: false,
 		UseWildcardsExpansion:      true,
+		MaxBufferSize:              defaultMaxBufferSize,
 		Log:                        testutil.Logger{},
 	}
 

